src/System.Windows.Extensions/src/System/Media/SoundPlayer.cs

```diff
@@ -0,0 +1,738 @@
+                int totalBytes = readBytes;
+
+                while (readBytes > 0)
+                {
+                    _currentPos += readBytes;
+                    if (_streamData.Length < _currentPos + BlockSize)
+                    {
+                        byte[] newData = new byte[_streamData.Length * 2];
+                        Array.Copy(_streamData, newData, _streamData.Length);
+                        _streamData = newData;
+                    }
+                    readBytes = _stream.Read(_streamData, _currentPos, BlockSize);
+                    totalBytes += readBytes;
+                }
+
+                _lastLoadException = null;
+            }
+            catch (Exception exception)
+            {
+                _lastLoadException = exception;
+            }
+
+            if (!_doesLoadAppearSynchronous)
+            {
+                // Post notification back to the UI thread.
+                _asyncOperation.PostOperationCompleted(
+                    _loadAsyncOperationCompleted,
+                    new AsyncCompletedEventArgs(_lastLoadException, false, null));
+            }
+            IsLoadCompleted = true;
+            _semaphore.Set();
+        }
+
+        private unsafe void ValidateSoundFile(string fileName)
+        {
+            IntPtr hMIO = Interop.WinMM.mmioOpen(fileName, IntPtr.Zero, Interop.WinMM.MMIO_READ | Interop.WinMM.MMIO_ALLOCBUF);
+            if (hMIO == IntPtr.Zero)
+            {
+                throw new FileNotFoundException(SR.SoundAPIFileDoesNotExist, _soundLocation);
+            }
+
+            try
+            {
+                Interop.WinMM.WAVEFORMATEX waveFormat = null;
+                var ckRIFF = new Interop.WinMM.MMCKINFO()
+                {
+                    fccType = mmioFOURCC('W', 'A', 'V', 'E')
+                };
+                var ck = new Interop.WinMM.MMCKINFO();
+                if (Interop.WinMM.mmioDescend(hMIO, ckRIFF, null, Interop.WinMM.MMIO_FINDRIFF) != 0)
+                {
+                    throw new InvalidOperationException(SR.Format(SR.SoundAPIInvalidWaveFile, _soundLocation));
+                }
+
+                while (Interop.WinMM.mmioDescend(hMIO, ck, ckRIFF, 0) == 0)
+                {
+                    if (ck.dwDataOffset + ck.cksize > ckRIFF.dwDataOffset + ckRIFF.cksize)
+                    {
+                        throw new InvalidOperationException(SR.SoundAPIInvalidWaveHeader);
+                    }
+
+                    if (ck.ckID == mmioFOURCC('f', 'm', 't', ' '))
+                    {
+                        if (waveFormat == null)
+                        {
+                            int dw = ck.cksize;
+                            if (dw < Marshal.SizeOf(typeof(Interop.WinMM.WAVEFORMATEX)))
+                            {
+                                dw = Marshal.SizeOf(typeof(Interop.WinMM.WAVEFORMATEX));
+                            }
```

Comment:
Seems like we'll want to clean up this marshaling stuff, change the type to a struct, use sizeof and avoid using PtrToStructure, etc.